### PR TITLE
More nullability annotations

### PR DESCRIPTION
**Pull request by mibac138**
Added nullability annotations to:
- Disposables
- Observers
and some others

### CODE REVIEW

**Review by vanniktech — CHANGES_REQUESTED** [review]
You probably don't want to mess up the imports.

File: src/main/java/io/reactivex/Single.java [hunk 1]
```diff
@@ -13,27 +13,42 @@
...
 import io.reactivex.exceptions.Exceptions;
 import io.reactivex.functions.*;
-import io.reactivex.internal.functions.*;
-import io.reactivex.internal.fuseable.*;
-import io.reactivex.internal.observers.*;
-import io.reactivex.internal.operators.completable.*;
-import io.reactivex.internal.operators.flowable.*;
-import io.reactivex.internal.operators.maybe.*;
-import io.reactivex.internal.operators.observable.*;
+import io.reactivex.internal.functions.Functions;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Single.java]: Please don't unroll star imports.

File: src/main/java/io/reactivex/disposables/ActionDisposable.java [hunk 2]
```diff
@@ -12,19 +12,20 @@
...
+import io.reactivex.annotations.NonNull;
 import io.reactivex.functions.Action;
 import io.reactivex.internal.util.ExceptionHelper;
 
 final class ActionDisposable extends ReferenceDisposable<Action> {
 
     private static final long serialVersionUID = -8219729196779211169L;
 
-    ActionDisposable(Action value) {
+    ActionDisposable(@NonNull Action value) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/ActionDisposable.java]: Non-public components don't really need such annotations.

File: src/main/java/io/reactivex/disposables/CompositeDisposable.java [hunk 3]
```diff
@@ -12,12 +12,16 @@
  */
 package io.reactivex.disposables;
 
-import java.util.*;
-
-import io.reactivex.exceptions.*;
+import io.reactivex.annotations.NonNull;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/CompositeDisposable.java]: Please don't unroll star imports.

File: src/main/java/io/reactivex/disposables/Disposables.java [hunk 4]
```diff
@@ -13,13 +13,14 @@
...
 
-import java.util.concurrent.Future;
-
-import org.reactivestreams.Subscription;
-
+import io.reactivex.annotations.NonNull;
 import io.reactivex.functions.Action;
 import io.reactivex.internal.disposables.EmptyDisposable;
-import io.reactivex.internal.functions.*;
+import io.reactivex.internal.functions.Functions;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/Disposables.java]: Please don't unroll star imports.

File: src/main/java/io/reactivex/disposables/FutureDisposable.java [hunk 5]
```diff
@@ -24,7 +26,7 @@
 
     private final boolean allowInterrupt;
 
-    FutureDisposable(Future<?> run, boolean allowInterrupt) {
+    FutureDisposable(@Nullable Future<?> run, boolean allowInterrupt) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/FutureDisposable.java]: Non-public components don't really need this annotation.

File: src/main/java/io/reactivex/disposables/ReferenceDisposable.java [hunk 6]
```diff
@@ -27,11 +28,11 @@
 
     private static final long serialVersionUID = 6537757548749041217L;
 
-    ReferenceDisposable(T value) {
+    ReferenceDisposable(@NonNull T value) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/ReferenceDisposable.java]: Non-public components don't need this annotation.

File: src/main/java/io/reactivex/disposables/RunnableDisposable.java [hunk 7]
```diff
@@ -12,19 +12,21 @@
...
+
 /**
  * A disposable container that manages a Runnable instance.
  */
 final class RunnableDisposable extends ReferenceDisposable<Runnable> {
 
     private static final long serialVersionUID = -8219729196779211169L;
 
-    RunnableDisposable(Runnable value) {
+    RunnableDisposable(@NonNull Runnable value) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/RunnableDisposable.java]: Non-public...

File: src/main/java/io/reactivex/disposables/SubscriptionDisposable.java [hunk 8]
```diff
@@ -21,12 +22,12 @@
 
     private static final long serialVersionUID = -707001650852963139L;
 
-    SubscriptionDisposable(Subscription value) {
+    SubscriptionDisposable(@NonNull Subscription value) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/disposables/SubscriptionDisposable.java]: Non-public...

File: src/main/java/io/reactivex/flowables/ConnectableFlowable.java [hunk 9]
```diff
@@ -113,6 +117,7 @@ public final Disposable connect() {
      *         when the specified number of Subscribers subscribe to it and calls the
      *         specified callback with the Subscription associated with the established connection
      */
+    @NonNull
     public Flowable<T> autoConnect(int numberOfSubscribers, Consumer<? super Disposable> connection) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/flowables/ConnectableFlowable.java]: NonNull could be added to the `connection` parameter as well as to `connect(Consumer)`.

**Comment by mibac138** [comment]
@vanniktech is it okay now?

File: src/main/java/io/reactivex/internal/disposables/DisposableHelper.java [hunk 10]
```diff
@@ -35,7 +37,7 @@
      * @param d the disposable to check
      * @return true if d is {@link #DISPOSED}
      */
-    public static boolean isDisposed(Disposable d) {
+    public static boolean isDisposed(@Nullable Disposable d) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/disposables/DisposableHelper.java]: For anything below `internal.*` I don't think annotations are necessary.

File: src/main/java/io/reactivex/observables/ConnectableObservable.java [hunk 11]
```diff
@@ -113,6 +117,7 @@ public final Disposable connect() {
      *         when the specified number of Subscribers subscribe to it and calls the
      *         specified callback with the Subscription associated with the established connection
      */
+    @NonNull
     public Observable<T> autoConnect(int numberOfSubscribers, Consumer<? super Disposable> connection) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observables/ConnectableObservable.java]: Non-null consumer and `connect(Consumer)` also needs annotation.

File: src/main/java/io/reactivex/observers/DisposableCompletableObserver.java [hunk 12]
```diff
@@ -50,10 +51,11 @@
  * </pre></code>
  */
 public abstract class DisposableCompletableObserver implements CompletableObserver, Disposable {
+    @NonNull
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/DisposableCompletableObserver.java]: I'm not sure non-public fields actually need such annotations.

File: src/main/java/io/reactivex/observers/ResourceSingleObserver.java [hunk 13]
```diff
@@ -76,9 +77,11 @@
  */
 public abstract class ResourceSingleObserver<T> implements SingleObserver<T>, Disposable {
     /** The active subscription. */
+    @NonNull
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/ResourceSingleObserver.java]: Unnecessary.

File: src/main/java/io/reactivex/observers/SafeObserver.java [hunk 14]
```diff
@@ -13,9 +13,13 @@
 package io.reactivex.observers;
 
 import io.reactivex.Observer;
+import io.reactivex.annotations.NonNull;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/SafeObserver.java]: Please don't unroll star imports.

File: src/main/java/io/reactivex/observers/SafeObserver.java [hunk 15]
```diff
@@ -26,8 +30,10 @@
  */
 public final class SafeObserver<T> implements Observer<T>, Disposable {
     /** The actual Subscriber. */
+    @NonNull
     final Observer<? super T> actual;
     /** The subscription. */
+    @Nullable
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/SafeObserver.java]: No need for it on non-public fields.

File: src/main/java/io/reactivex/parallel/ParallelFlowable.java [hunk 16]
```diff
@@ -13,20 +13,24 @@
 
 package io.reactivex.parallel;
 
-import java.util.*;
-import java.util.concurrent.Callable;
-
-import org.reactivestreams.*;
-
-import io.reactivex.*;
+import io.reactivex.Flowable;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/parallel/ParallelFlowable.java]: Please don't unroll star imports.

File: src/main/java/io/reactivex/processors/AsyncProcessor.java [hunk 17]
```diff
@@ -12,14 +12,13 @@
  */
 package io.reactivex.processors;
 
-import io.reactivex.annotations.CheckReturnValue;
-import java.util.Arrays;
-import java.util.concurrent.atomic.AtomicReference;
-
-import org.reactivestreams.*;
-
+import io.reactivex.annotations.*;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/processors/AsyncProcessor.java]: What IDE are you using that reorders imports?

File: src/main/java/io/reactivex/processors/AsyncProcessor.java [hunk 18]
```diff
@@ -36,7 +35,8 @@
 
     @SuppressWarnings("rawtypes")
     static final AsyncSubscription[] TERMINATED = new AsyncSubscription[0];
-
+    
+    @NonNull
```
Comment (akarnokd) [inline src/main/java/io/reactivex/processors/AsyncProcessor.java]: Non-public, not necessary.

File: src/main/java/io/reactivex/schedulers/Timed.java [hunk 19]
```diff
@@ -13,18 +13,21 @@
...
 
+import java.util.concurrent.TimeUnit;
+
 /**
  * Holds onto a value along with time information.
  *
  * @param <T> the value type
  */
 public final class Timed<T> {
+    @Nullable
```
Comment (akarnokd) [inline src/main/java/io/reactivex/schedulers/Timed.java]: Not necessary.

File: src/main/java/io/reactivex/schedulers/Timed.java [hunk 20]
```diff
@@ -34,7 +37,7 @@
      * @param unit the time unit, not null
      * @throws NullPointerException if unit is null
      */
-    public Timed(T value, long time, TimeUnit unit) {
+    public Timed(@Nullable T value, long time, @NonNull TimeUnit unit) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/schedulers/Timed.java]: Why is this nullable?

**Review by akarnokd — CHANGES_REQUESTED** [review]
Some of the non-public components don't need annotation.

File: src/main/java/io/reactivex/processors/AsyncProcessor.java (hunk 17 above)
Comment (mibac138) [reply src/main/java/io/reactivex/processors/AsyncProcessor.java]: IntelliJ IDEA

File: src/main/java/io/reactivex/exceptions/CompositeException.java [hunk 21]
```diff
@@ -47,7 +49,7 @@
      *
      * @throws IllegalArgumentException if <code>exceptions</code> is empty.
      */
-    public CompositeException(Throwable... exceptions) {
+    public CompositeException(@Nullable Throwable... exceptions) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/CompositeException.java]: This should be `@NonNull`.

File: src/main/java/io/reactivex/exceptions/CompositeException.java [hunk 22]
```diff
@@ -59,7 +61,7 @@ public CompositeException(Throwable... exceptions) {
      *
      * @throws IllegalArgumentException if <code>errors</code> is empty.
      */
-    public CompositeException(Iterable<? extends Throwable> errors) {
+    public CompositeException(@Nullable Iterable<? extends Throwable> errors) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/CompositeException.java]: This should be `@NonNull`.

File: src/main/java/io/reactivex/exceptions/Exceptions.java [hunk 23]
```diff
@@ -32,7 +33,8 @@ private Exceptions() {
      * @return because {@code propagate} itself throws an exception or error, this is a sort of phantom return
      *         value; {@code propagate} does not actually return anything
      */
-    public static RuntimeException propagate(Throwable t) {
+    @NonNull
+    public static RuntimeException propagate(@Nullable Throwable t) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/Exceptions.java]: The argument should be `@NonNull`.

File: src/main/java/io/reactivex/exceptions/Exceptions.java [hunk 24]
```diff
@@ -61,7 +63,7 @@ public static RuntimeException propagate(Throwable t) {
      *         the {@code Throwable} to test and perhaps throw
      * @see <a href="https://github.com/ReactiveX/RxJava/issues/748#issuecomment-32471495">RxJava: StackOverflowError is swallowed (Issue #748)</a>
      */
-    public static void throwIfFatal(Throwable t) {
+    public static void throwIfFatal(@Nullable Throwable t) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/Exceptions.java]: The argument should be `@NonNull`.

File: src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java [hunk 25]
```diff
@@ -35,7 +35,7 @@
      * @param e
      *          the {@code Throwable} to signal; if null, a NullPointerException is constructed
      */
-    public OnErrorNotImplementedException(String message, Throwable e) {
+    public OnErrorNotImplementedException(String message, @Nullable Throwable e) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java]: The `e` argument should be `@NonNull`.

File: src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java [hunk 26]
```diff
@@ -47,7 +47,7 @@ public OnErrorNotImplementedException(String message, Throwable e) {
      * @param e
      *          the {@code Throwable} to signal; if null, a NullPointerException is constructed
      */
-    public OnErrorNotImplementedException(Throwable e) {
+    public OnErrorNotImplementedException(@Nullable Throwable e) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java]: The argument should be non-null. It's just a safeguard to make sure a null is recorded as NPE.

File: src/main/java/io/reactivex/flowables/GroupedFlowable.java [hunk 27]
```diff
@@ -31,13 +32,14 @@
  */
 public abstract class GroupedFlowable<K, T> extends Flowable<T> {
 
+    @Nullable
```
Comment (akarnokd) [inline src/main/java/io/reactivex/flowables/GroupedFlowable.java]: Nullable annotation on package-private fields are unnecessary.

File: src/main/java/io/reactivex/internal/disposables/DisposableHelper.java [hunk 28]
```diff
@@ -13,13 +13,13 @@
...
 package io.reactivex.internal.disposables;
 
-import java.util.concurrent.atomic.AtomicReference;
-
 import io.reactivex.disposables.Disposable;
 import io.reactivex.exceptions.ProtocolViolationException;
 import io.reactivex.internal.functions.ObjectHelper;
 import io.reactivex.plugins.RxJavaPlugins;
 
+import java.util.concurrent.atomic.AtomicReference;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/disposables/DisposableHelper.java]: Please restore the original file if there is no actual (code) change.

File: src/main/java/io/reactivex/observables/GroupedObservable.java [hunk 29]
```diff
@@ -31,13 +32,14 @@
  */
 public abstract class GroupedObservable<K, T> extends Observable<T> {
 
+    @Nullable
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observables/GroupedObservable.java]: No need to annotate this.

File: src/main/java/io/reactivex/observers/SerializedObserver.java [hunk 30]
```diff
@@ -34,7 +35,7 @@
     final boolean delayError;
 
     static final int QUEUE_LINK_SIZE = 4;
-
+
```
Comment (akarnokd) [inline src/main/java/io/reactivex/observers/SerializedObserver.java]: Whitespace added?

File: src/main/java/io/reactivex/processors/AsyncProcessor.java [hunk 31]
```diff
@@ -36,7 +35,7 @@
 
     @SuppressWarnings("rawtypes")
     static final AsyncSubscription[] TERMINATED = new AsyncSubscription[0];
-
+
```
Comment (akarnokd) [inline src/main/java/io/reactivex/processors/AsyncProcessor.java]: Whitespace added?

File: src/main/java/io/reactivex/schedulers/Timed.java [hunk 32]
```diff
@@ -44,6 +45,7 @@ public Timed(T value, long time, TimeUnit unit) {
      * Returns the contained value.
      * @return the contained value
      */
+    @Nullable
```
Comment (akarnokd) [inline src/main/java/io/reactivex/schedulers/Timed.java]: This should be `@NonNull`.

File: src/main/java/io/reactivex/subjects/SingleSubject.java [hunk 33]
```diff
@@ -34,7 +34,7 @@
  */
 @Experimental
 public final class SingleSubject<T> extends Single<T> implements SingleObserver<T> {
-
+
```
Comment (akarnokd) [inline src/main/java/io/reactivex/subjects/SingleSubject.java]: Whitespace?

**Review by akarnokd — COMMENTED** [review]
A couple of changes are still needed.

**Review by akarnokd — APPROVED** [review]

File: src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java [hunk 34]
```diff
@@ -35,7 +35,7 @@
      * @param e
      *          the {@code Throwable} to signal; if null, a NullPointerException is constructed
      */
-    public OnErrorNotImplementedException(String message, Throwable e) {
+    public OnErrorNotImplementedException(String message, @NonNull Throwable e) {
```
Comment (PaulWoitaschek) [inline src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java:38]: Looks like this should be Nullable

File: src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java [hunk 35]
```diff
@@ -47,7 +47,7 @@ public OnErrorNotImplementedException(String message, Throwable e) {
      * @param e
      *          the {@code Throwable} to signal; if null, a NullPointerException is constructed
      */
-    public OnErrorNotImplementedException(Throwable e) {
+    public OnErrorNotImplementedException(@NonNull Throwable e) {
```
Comment (PaulWoitaschek) [inline src/main/java/io/reactivex/exceptions/OnErrorNotImplementedException.java:50]: Nullable too.

File: src/main/java/io/reactivex/exceptions/CompositeException.java [hunk 36]
```diff
@@ -47,7 +49,7 @@
      *
      * @throws IllegalArgumentException if <code>exceptions</code> is empty.
      */
-    public CompositeException(Throwable... exceptions) {
+    public CompositeException(@NonNull Throwable... exceptions) {
```
Comment (PaulWoitaschek) [inline src/main/java/io/reactivex/exceptions/CompositeException.java:52]: Nullable instead? The code handles it explicitly.

**Pull request merged**